src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+                action();
+            }
+            catch
+            {
+                if (!IsErrorState(this.state))
+                {
+                    this.SetState(BatchWriterState.Error);
+                }
+
+                throw;
+            }
         }
 
         /// <summary>
-        /// Ends a batch - implementation of the actual functionality.
+        /// Wrapper method to validate state transition with optional customized validation.
         /// </summary>
-        private void WriteEndBatchImplementation()
+        /// <param name="newState">Teh new writer state to transition into.</param>
+        /// <param name="customizedValidationAction">Optional validation action.</param>
+        internal void ValidateTransition(BatchWriterState newState, Action customizedValidationAction)
         {
-            Debug.Assert(
-                this.batchStartBoundaryWritten || this.CurrentOperationMessage == null,
-                "If not batch boundary was written we must not have an active message.");
-
-            // write pending message data (headers, response line) for a previously unclosed message/request
-            this.WritePendingMessageData(true);
-
-            this.SetState(BatchWriterState.BatchCompleted);
-
-            // write the end boundary for the batch
-            ODataBatchWriterUtils.WriteEndBoundary(this.rawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
+            if (customizedValidationAction != null)
+            {
+                customizedValidationAction();
+            }
 
-            // For compatibility with WCF DS we write a newline after the end batch boundary.
-            // Technically it's not needed, but it doesn't violate anything either.
-            this.rawOutputContext.TextWriter.WriteLine();
+            CommonTransitionValidation(newState);
         }
 
         /// <summary>
-        /// Verifies that calling WriteStartChangeset is valid.
+        /// Sets the 'Error' state and then throws an ODataException with the specified error message.
         /// </summary>
-        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanWriteStartChangeset(bool synchronousCall)
+        /// <param name="errorMessage">The error message for the exception.</param>
+        internal void ThrowODataException(string errorMessage)
         {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
+            this.SetState(BatchWriterState.Error);
+            throw new ODataException(errorMessage);
         }
 
         /// <summary>
-        /// Starts a new changeset - implementation of the actual functionality.
+        /// Increases the size of the current batch message; throws if the allowed limit is exceeded.
         /// </summary>
-        private void WriteStartChangesetImplementation()
+        internal void IncreaseBatchSize()

```

Comment:
>IncreaseBatchSize [](start = 22, length = 17)

Can IncreaseBatchSize, IncreaseChangesetSize, and ResetChangesetSize all be made private, and called in CreateOperationRequestMessage (for IncreaseBatchSize/IncreaseChangesetSize) and WriteStartChangeset (for ResetChangeset) in the base class, rather than in CreateOperationRequestMessageImplementation (for IncreaseBatchSize/IncreaseChangesetSize)  and WriteStartChangesetImplementation (for ResetChangesetSize) in the derived class?

Reply:
I doubt that. The IncreaseBatchSize/IncreaseChangeSetSize needs to be invoked in derived class since it needs changesetBoundary that is not in base class ODataBatchWriter. (see ODataMultipartMixedBatchWriter.CreateOperationRequestMessageImplementation)

---
In reply to: [140933241](https://github.com/OData/odata.net/pull/960#discussion_r140933241) [](ancestors = 140933241)

Reply:
Logically, if we refactor correctly, this should all be in the ReadImplementation base class.

---
In reply to: [142589288](https://github.com/OData/odata.net/pull/960#discussion_r142589288) [](ancestors = 142589288,140933241)

Reply:
This belongs to Writer's CreateOperationRequestMessage instead.
Added abstract IsInsideSubBatch() in base class, so that the base class can know whether the batch size or the changeset size is the counter that needs to be incremented.

---
In reply to: [142823974](https://github.com/OData/odata.net/pull/960#discussion_r142823974) [](ancestors = 142823974,142589288,140933241)

Reply:
Making these private looks good in latest iteration. I created a separate comment for the new IsInsideSubBatch(), which should not be required (base class has all of the information it needs to know if it is in a changeset or not).

---
In reply to: [143320362](https://github.com/OData/odata.net/pull/960#discussion_r143320362) [](ancestors = 143320362,142823974,142589288,140933241)